tests/test_extract_contained_procedures.py

```diff
@@ -0,0 +1,594 @@
+    the derived type itself, that is, 'a', is introduced as an the argument in the transformation.
+    """
+    fcode = """
+        subroutine outer()
+            implicit none
+            use types, only: my_type, your_type
+            type(my_type) :: xtyp
+            type(your_type) :: ytyp
+            call inner()
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                xtyp%a = 40
+                ytyp%val%b = 10.0
+                z = y + ytyp%something_else
+            end subroutine inner
+        end subroutine outer
+    """
+    src = Sourcefile.from_source(fcode, frontend = frontend)
+    routines = extract_contained_procedures(src.routines[0])
+    outer = src.routines[0]
+    inner = routines[0]
+    assert 'xtyp' in inner.arguments
+    assert 'ytyp' in inner.arguments
+
+    call = FindNodes(CallStatement).visit(outer.body)[0]
+    kwargdict = dict(call.kwarguments)
+    assert kwargdict['xtyp'] == 'xtyp'
+    assert kwargdict['ytyp'] == 'ytyp'
+
+    imports = FindNodes(Import).visit(inner.spec)
+    modules = set()
+    symbols = set()
+    for imp in imports:
+        modules.add(imp.module)
+        for sym in imp.symbols:
+            symbols.add(sym)
+    assert "types" in modules
+    assert len(modules) == 1
+    assert "my_type" in symbols
+    assert "your_type" in symbols
+    assert len(symbols) == 2
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_extract_contained_procedures_intent(frontend):
+    """
+    This test is just to document the current behaviour: when a global is
+    introduced as an argument to the extracted contained procedure,
+    its intent will be 'inout', unless the intent is specified in the parent procedure.
+    """
+    fcode = """
+        subroutine outer(v)
+            implicit none
+            integer, intent(in) :: v
+            integer, intent(out) :: p
+            integer :: x(3)
+            x = 4
+            call inner()
+            p = 400
+            contains
+            subroutine inner()
+                integer :: y
+                integer :: z
+                y = 1
+                z = x(1) + v + y + p
+            end subroutine inner
+        end subroutine outer
+    """
+    src = Sourcefile.from_source(fcode, frontend = frontend)
+    routines = extract_contained_procedures(src.routines[0])
+    assert len(routines) == 1
+    outer = src.routines[0]
+    inner = routines[0]
+    assert inner.variable_map['v'].type.intent == "in"
+    assert inner.variable_map['x'].type.intent == "inout"
+    assert inner.variable_map['p'].type.intent == "out"
+
+    # Also check that the intents don't change in the parent.
+    assert outer.variable_map['v'].type.intent == "in"
+    assert outer.variable_map['x'].type.intent is None
+    assert outer.variable_map['p'].type.intent == "out"
+
+@pytest.mark.parametrize('frontend', available_frontends())
```

Comment:
For this test we have to mark OMNI as xfail (which is fine, it's a limitation of the frontend) because it is refusing to parse code with undefined symbols:
```suggestion
@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI, 'Parser fails on undefined symbols')]))
```